gramps/gen/display/name.py

```diff
@@ -1288,18 +1287,17 @@ def ifNotEmpty(str,p,s):
             ",".join(param),
         )
         try:
-            exec(s) in globals(), locals()
-            return locals()["fn"]
-        except:
+            result = {}
+            exec(s, globals(), result)
+            return result["fn"]
+        except SyntaxError:
```

Comment:
Wrong format strings produce TypeErrors:

```pycon
>>> '' % 'a'
Traceback (most recent call last):
  File "<python-input-0>", line 1, in <module>
    '' % 'a'
    ~~~^~~~~
TypeError: not all arguments converted during string formatting
>>> '%x' % 'a'
Traceback (most recent call last):
  File "<python-input-1>", line 1, in <module>
    '%x' % 'a'
    ~~~~~^~~~~
TypeError: %x format: an integer is required, not str
>>> '%s %s' % 'a'
Traceback (most recent call last):
  File "<python-input-2>", line 1, in <module>
    '%s %s' % 'a'
    ~~~~~~~~^~~~~
TypeError: not enough arguments for format string
```

Reply:
Format strings are only evaluated on the line defining `s` earlier, or within `fn` itself, but not at the outer level at which the `exec` is run.

All the examples you've given pass when wrapped in `def fn():`.

Note, `fn` is not called here, but cached for use in the callers of this function.

Reply:
Alright then, sorry for the noise.